esp/services/ws_workunits/ws_wuresult.cpp

```diff
@@ -0,0 +1,441 @@
+    }
+    else if (outFormat == WUResultOutJSON)
+    {
+        context->setResponseFormat(ESPSerializationJSON);
+        context->addCustomerHeader("Content-disposition", "attachment;filename=WUResult.json");
+    }
+    else if (outFormat == WUResultOutXML)
+    {
+        context->setResponseFormat(ESPSerializationXML);
+        context->addCustomerHeader("Content-disposition", "attachment;filename=WUResult.xml");
+    }
+    else
+    {
+        StringBuffer headerStr("attachment;filename=WUResult.xml");
+        if (outFormat == WUResultOutGZIP)
+            headerStr.append(".gz");
+        else
+            headerStr.append(".zip");
+        context->addCustomerHeader("Content-disposition", headerStr.str());
+    }
+}
+
+void CWsWuResultOutHelper::startStreaming()
+{
+    if (resultIOS)
+        resultIOS->write(strlen(wuResultXMLStartStr), wuResultXMLStartStr);
+    else
+    {
+        response->setStatus(HTTP_STATUS_OK);
+        response->startSend();
+        if (outFormat == WUResultOutXML)
+            response->sendChunk(wuResultXMLStartStr);
+    }
+}
+
+void CWsWuResultOutHelper::finalXMLStreaming()
+{
+    if (resultIOS)
+        resultIOS->write(strlen(wuResultXMLEndStr), wuResultXMLEndStr);
+    else
+        response->sendFinalChunk(wuResultXMLEndStr);
+}
+
+//Based on CWsWorkunitsEx::getWsWuResult
+void CWsWuResultOutHelper::getWsWuResult(const char* resultName, const char* logicalName, unsigned sequence)
+{
+    Owned<IWorkUnitFactory> factory = getWorkUnitFactory(context->querySecManager(), context->queryUser());
+    Owned<IConstWorkUnit> cw = factory->openWorkUnit(wuid);
+    if (!cw)
+        throw makeStringExceptionV(ECLWATCH_CANNOT_OPEN_WORKUNIT, "Cannot open workunit %s.", wuid.str());
+
+    Owned<IConstWUResult> result;
+    if (notEmpty(resultName))
+        result.setown(cw->getResultByName(resultName));
+    else if (notEmpty(logicalName))
+    {
+        Owned<IConstWUResultIterator> it = &cw->getResults();
+        ForEach(*it)
+        {
+            IConstWUResult& r = it->query();
+            SCMStringBuffer filename;
+            if (strieq(r.getResultLogicalName(filename).str(), logicalName))
+            {
+                result.setown(LINK(&r));
+                break;
+            }
+        }
+    }
+    else
+        result.setown(cw->getResultBySequence(sequence));
+    
+    if (!result)
+        throw makeStringException(ECLWATCH_CANNOT_GET_WU_RESULT, "Cannot open the workunit result.");
+
+    SCMStringBuffer logicalNameBuf;
+    result->getResultLogicalName(logicalNameBuf);
+
+    Owned<INewResultSet> rs;
+    Owned<IResultSetFactory> resultSetFactory = getSecResultSetFactory(context->querySecManager(),
+        context->queryUser(), context->queryUserId(), context->queryPassword());
+    if (logicalNameBuf.length())
+    {
+        rs.setown(resultSetFactory->createNewFileResultSet(logicalNameBuf.str(), cw->queryClusterName())); //MORE is this wrong cluster?
```

Comment:
The comment is at least 6 years old... would we know by now if it was the wrong cluster?

Reply:
We still do not know. But, we try it because there is no other option.